### PR TITLE
task is done

### PR DESCRIPTION
**Pull request by yermilov-m**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,131 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int DEFAULT_SIZE = 10;
+    private T[] array;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:8]:
```suggestion
    private T[] elements;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,131 @@
...
+    private int size;
+    private int maxSize;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:10]:
```suggestion
```
maxSize дублирует значение elements.length

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,131 @@
...
 
+    public ArrayList() {
+        size = 0;
+        maxSize = DEFAULT_SIZE;
+        array = (T[]) (new Object[maxSize]);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:15]:
```suggestion
        array = (T[]) (new Object[maxSize]);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,131 @@
...
-
+        if (size >= maxSize) {
+            array = getExpanded();
+        }
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("You can't add an element at that index");
+        } else if (index == size) {
+            array[index] = value;
+        } else {
+            array = insertAndCopy(value, index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:32]: Copying is part of insert logic. No need to have this word in the method name.
```suggestion
            array = insertElement(value, index);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,131 @@
...
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:40]:
```suggestion
            add(list.get(i));
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,131 @@
...
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("There is no element with such index");
+        } else {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:48]: image

**Review by Rommelua — CHANGES_REQUESTED** [review]
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html

Before you create PR with your homework, you need to go through the checklist under the task and correct all the points described there. The mentor will not check the work until the checklist points are corrected

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,120 @@
...
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("There is no element with such index");
+        } else {
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:45]: move this check to a separate method checkIndex(index)

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,120 @@
...
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T oldValue = get(index);
+        elements = removeAndCopy(index);
+        size--;
+        return oldValue;
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:62]: use System.arraycopy to move all the elements 1 position to the left

hint - you can transfer elements from and to the same array

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,120 @@
...
+        if (size >= elements.length) {
+            elements = getExpanded();
+        }
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("You can't add an element at that index");
+        } else if (index == size) {
+            elements[index] = value;
+        } else {
+            elements = insertElement(value, index);
+        }
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:30]: use System.arraycopy to move all the elements 1 position to the right

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,120 @@
...
+            return elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (get(index) != null) {
+            elements[index] = value;
+        }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:54]: А если в ячейке лежит null то нам уже не нужно делать set? 
[15, -2, null, 28]
set(10, 2) не отработает?

**Review by Rommelua — CHANGES_REQUESTED** [review]
You should fix (or answer to) all mentor's comments before re-request new review. Be attentive and do not hesitate to contact us in Slack if you do not understand something. 
Do not re-request new review if discussion in Slack is goin on)

**Comment by yermilov-m** [comment]
Sorry for spamming review requests. Meesa GitHub noob

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,104 @@
...
+    private static final int DEFAULT_SIZE = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) (new Object[DEFAULT_SIZE]);
+    }
 
+    public void add(T value) {
+        add(value, size);
```
Comment (andriy-shulepa) [inline src/main/java/core/basesyntax/ArrayList.java:16]: it's good that you try to reuse code. however, adding an element to the end of the list does not require copying the array to free space for a new element. So, your code is working but less performant

**Review by andriy-shulepa — APPROVED** [review]

File: src/main/java/core/basesyntax/ArrayList.java (hunk 11 above)
Comment (yermilov-m) [reply src/main/java/core/basesyntax/ArrayList.java:16]: You're right. Code edited

**Pull request closed without merging**